### PR TITLE
Test load-images=yes and load-images=no.

### PR DESCRIPTION
**Pull request by mepard**
Tests the function of the --load-images command line option. When compiled with --debug, demonstrates the leak described in #12903.

Replaces previous pull request #13919.

### CODE REVIEW

File: test/lib/www/load-images.html [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<html>
+<body>
+    <img src="logo.png"/>
+    <img src="phantomjs.png"/>
```
Comment (zackw) [inline test/lib/www/load-images.html]: Please don't use XMLisms (``) in a file that isn't XML.  (Please *do* add `<!doctype html>` at the top of this file.)

File: test/module/webpage/load-images-no.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        page.release();
+    }
+    
+    page.open(url, this.step_func_done (function (status) {
+        assert_equals(status, "success");
+        checkImageCounts ();
+    }));
+}
+
+async_test(do_test, "load a page two images with --load-images=no");
```
Comment (zackw) [inline test/module/webpage/load-images-no.js]: Is there a reason `do_test` is not an anonymous function inside the call to `async_test`?

File: test/module/webpage/load-images-no.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
+//! phantomjs: --load-images=no
+
+// Compiled for debug, this tests for leaks when not loading images.
+//
```
Comment (zackw) [inline test/module/webpage/load-images-no.js]: Is the line with just `//` on it an editing mistake?

File: test/module/webpage/load-images-no.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    });
+
+    page.onResourceReceived = this.step_func(function (resource) {
+        if (resource.stage === 'end' && /\.png$/.test(resource.url)) ++numImagesLoaded;
+    });
+
+    function checkImageCounts ()
+    {
+        var imageCounts = page.evaluate (function() {    
+            "use strict";
```
Comment (zackw) [inline test/module/webpage/load-images-no.js]: If you're going to `"use strict"` please do it for the entire file.

File: test/module/webpage/load-images-no.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            {
+                var images = document.getElementsByTagName('img');
+                var imageCounts = {};
+                imageCounts.numImages = images.length;
+                imageCounts.numComplete = Array.prototype.reduce.call(images, function(n, image) { return n + (image.complete? 1 : 0); }, 0);
+                return imageCounts;
+            }
+            catch(err)
+            {
+                return undefined;
```
Comment (zackw) [inline test/module/webpage/load-images-no.js]: What exceptions might get thrown in this context?

File: test/module/webpage/load-images-no.js (hunk 2 above)
Comment (mepard) [inline test/module/webpage/load-images-no.js]: These evolved from pjs-10690.js. Other than that, no reason at all. I personally wouldn't make it anonymous, though. I've gotten in the habit of naming all functions to improve traces (particularly in casperjs errors).

File: test/lib/www/load-images.html (hunk 1 above)
Comment (mepard) [inline test/lib/www/load-images.html]: I copied logo.html and didn't give it much thought.

File: test/module/webpage/load-images-no.js (hunk 4 above)
Comment (mepard) [inline test/module/webpage/load-images-no.js]: Does that work when it's inside evaluate?

File: test/module/webpage/load-images-no.js (hunk 5 above)
Comment (mepard) [inline test/module/webpage/load-images-no.js]: Don't know, but I've gotten into the habit of making sure problems in evaluate show up outside in some fashion.

File: test/module/webpage/load-images-no.js (hunk 4 above)
Comment (zackw) [inline test/module/webpage/load-images-no.js]: Oh, good point, it probably doesn't.  I think we serialize to text and recompile in the page context.  Leave that there and add one for the file, then?

Comment (mepard) [inline test/module/webpage/load-images-no.js]: Sounds good.

File: test/module/webpage/load-images-no.js (hunk 3 above)
Comment (mepard) [inline test/module/webpage/load-images-no.js]: Yup. Sorry.

File: test/module/webpage/load-images-no.js (hunk 5 above)
Comment (zackw) [inline test/module/webpage/load-images-no.js]: I'm not sure what happens if an exception is thrown and not caught inside `evaluate` -- but this way, the exception is swallowed and we just have `undefined` in a comparison later, which is safe (`assert_*` use strict equality) but might be baffling.  I guess what I want is to make sure that any potential exception gets logged somehow.

File: test/module/webpage/load-images-no.js (hunk 2 above)
Comment (zackw) [inline test/module/webpage/load-images-no.js]: I was mainly just curious; most of the tests use anon functions for this.

File: test/lib/www/load-images.html (hunk 1 above)
Comment (zackw) [inline test/lib/www/load-images.html]: Yeah that file could probably stand to have the dust blown off it a little.  No worries.

File: test/module/webpage/load-images-no.js (hunk 5 above)
Comment (mepard) [inline test/module/webpage/load-images-no.js]: Logging inside evaluate won't surface unless you use onConsoleMessage.

I played around with what happens when I intentionally throw an exception. In this case, it's not the return value that gets checked, but returnvalue.someField. This results in e.g.

	module/webpage/load-images-yes: FAIL
	 FAIL: load a page two images with --load-images=yes
	  null is not an object (evaluating 'imageCounts.numImages')
	  checkImageCounts@phantomjs://code/load-images-yes.js:38:34
	  pageOpened@phantomjs://code/load-images-yes.js:46:25

If I remove the try/catch from evaluate, the exception does get reported, but it's pretty strange:

	module/webpage/load-images-yes: ERROR
	ERROR: neither a test nor a diagnostic
	  Error: oops
	 FAIL: test 1 did not report status

Error: oops is the exception I added, but in most cases you wouldn't see the explicit throw in the test code. The other way seems clearer to me.

**Comment by mepard** [comment]
https://github.com/Vitallium/qtwebkit/pull/8

**Comment by ariya** [comment]
Excellent fix! Please go ahead and squash it into one commit and I'll land it. Thanks!

**Comment by Vitallium** [comment]
Squashed and landed in https://github.com/ariya/phantomjs/commit/4ef5f2760028f8f0c341bbad3915937587d93b3f

**Pull request closed without merging**